preview/Win7Msix/Win7MSIXInstallerLib/PackageInfo.hpp

```diff
@@ -0,0 +1,114 @@
+        unsigned long long m_version = 0;
+        std::wstring m_publisher;
+        std::wstring m_publisherName;
+        std::wstring m_relativeLogoPath;
+
+        /// PackageReader and payloadFiles are available on Add, but not Remove because it's created off the original package itself which is no longer available once it's been installed.
+        ComPtr<IAppxPackageReader> m_packageReader;
+        unsigned int m_numberOfPayloadFiles = 0;
+
+
+        /// Sets the executable path by reading it from the manifest element
+        HRESULT SetExecutableAndAppIdFromManifestElement(IMsixElement * element);
+
+        /// Sets the display name by reading it from the manifest element
+        HRESULT SetDisplayNameFromManifestElement(IMsixElement * element);
+    public:
+        std::wstring GetPackageFullName() { return m_packageFullName; }
+        std::wstring GetRelativeExecutableFilePath() { return m_relativeExecutableFilePath; }
+        std::wstring GetDisplayName() { return m_displayName; }
+        std::wstring GetId() { return m_appUserModelId; }
+        unsigned long long GetVersion() { return m_version; }
+        std::wstring GetPublisher() { return m_publisher; }
+        std::wstring GetPublisherDisplayName() { return m_publisherName; }
+
+    protected:
+        PackageInfoBase() {}
+
+        /// Sets the manifest reader, and other fields derived from the manifest
+        /// Specifically, packageFullName, packageDirectoryPath, executableFilePath, displayname, version and publisher.
+        ///
+        /// @param manifestReader - manifestReader to set
+        HRESULT SetManifestReader(IAppxManifestReader* manifestReader);
+    public:
+
+        /// When made from manifest reader, it won't have PackageReader available. 
+        bool HasPackageReader() { return (m_packageReader.Get() != nullptr); };
+        IAppxPackageReader * GetPackageReader() { return m_packageReader.Get(); }
+        unsigned int GetNumberOfPayloadFiles() { return m_numberOfPayloadFiles; }
+
+        // Getters
+        IAppxManifestReader * GetManifestReader() { return m_manifestReader.Get(); }
+        /// This is meant only to be called when deleting the manifest file; the reader needs to first be released so it can be deleted
+        void ReleaseManifest() { m_manifestReader.Release(); }
+    };
+
+
+    class PackageInfo : public PackageInfoBase, IPackageInfo
+    {
+    public:
+        PackageInfo() :PackageInfoBase() {}
+
+        std::wstring GetPackageFullName() { return m_packageFullName; }
+        std::wstring GetRelativeExecutableFilePath() { return m_relativeExecutableFilePath; }
+        std::wstring GetDisplayName() { return m_displayName; }
+        std::wstring GetId() { return m_appUserModelId; }
+        unsigned long long GetVersion() { return m_version; }
+        std::wstring GetPublisher() { return m_publisher; }
+        std::wstring GetPublisherDisplayName() { return m_publisherName; }

```

Comment:
these are the same as the base class right?

Reply:
Yes, but we can't remove them, the interfaces IPackageInfo and IInstalledPackageInfo (Diamond interface inheritance problem)